File: 202403/tests/FM_03_gibbs_julia.jl
using Random
using Distributions

# update rules

function calc_q_init(
    x::Array{Float64,2},
    v::Array{Float64,2}
)::Array{Float64,2}
    # x: (D, N)
    # v: (N, K)
    return x[:, :] * v[:, :] # (D, K)
end

function calc_dq(
    i::Int,
    x::Array{Float64,2},
    v_ik_new::Float64,
    v_ik_old::Float64
)::Array{Float64,1}
    # v_ik_new: float
    # v: (N, K)
    # x: (D, N)
    return (v_ik_new - v_ik_old) .* x[:, i] # (D)
end

function calc_df(
    x_theta::Array{Float64,1},
    param_new::Float64,
    param_old::Float64
)
    return (param_new - param_old) * x_theta
end

function calc_xy_b(
    f::Array{Float64,1},
    b::Float64,
    x_data::Array{Float64,2},
    y_data::Array{Float64,1}
)
    # x_data: (D, N)
    # y_data: (D)
    x_b = ones(size(x_data, 1))
    y_b = y_data - (f - b * x_b)
    return x_b, y_b
end

function calc_xy_w(
    f::Array{Float64,1},
    w::Array{Float64,1},
    x_data::Array{Float64,2},
    y_data::Array{Float64,1},
    i::Int
)
    # x_data: (D, N)
    # y_data: (D)
    x_w = x_data[:, i]
    y_w = y_data - (f - x_w .* w[i])
    return x_w, y_w
end

function calc_xy_v(
    f::Array{Float64,1},
    q::Array{Float64,2},
    v::Array{Float64,2},
    x_data::Array{Float64,2},
    y_data::Array{Float64,1},
    i::Int,
    k::Int
)
    # x_data: (D, N)
    # y_data: (D)
    x_v = x_data[:, i] .* (q[:, k] - x_data[:, i] .* v[i, k])
    y_v = y_data - (f - x_v .* v[i, k])
    return x_v, y_v
end

# samplers
function sample_from_inverse_gamma(
    a::Float64,
    b::Float64,
    seed::Union{Int, Nothing}=nothing
)
    if !isnothing(seed)
        Random.seed!(seed)
    end
    return 1 / rand(Gamma(a, 1 / b))
end

function sample_posterior_sigma2_noise(
    x_theta::Array{Float64,1},
    y_theta::Array{Float64,1},
    theta::Float64,
    a_noise::Float64,
    b_noise::Float64,
    seed::Union{Int, Nothing}=nothing
)
    # x_theta: (D)
    # y_theta: (D)
    D = length(x_theta)
    a_noise_post = a_noise + D / 2
    b_noise_post = b_noise + 0.5 * sum((y_theta - x_theta .* theta).^2)
    return sample_from_inverse_gamma(a_noise_post, b_noise_post, seed)
end

function sample_posterior_theta(
    x_theta::Array{Float64,1},
    y_theta::Array{Float64,1},
    sigma2_noise::Float64,
    mu_theta::Float64,
    sigma2_theta::Float64,
    seed::Union{Int, Nothing}=nothing
)
    if !isnothing(seed)
        Random.seed!(seed)
    end
    sigma2_theta_post = 1 / (sum(x_theta.^2) / sigma2_noise + 1 / sigma2_theta)
    mu_theta_post = sigma2_theta_post * (sum(x_theta .* y_theta) / sigma2_noise + mu_theta / sigma2_theta)
    return rand(Normal(mu_theta_post, sqrt(sigma2_theta_post)))
end

function sample_posterior_mu_theta_and_sigma2_theta(
    theta::Array{Float64,1},
    sigma2_theta::Float64,
    m_theta::Float64,
    l_theta::Float64,
    a_theta::Float64,
    b_theta::Float64,
    seed::Union{Int, Nothing}=nothing
)
    # theta: (N)
    if !isnothing(seed)
        Random.seed!(seed)
    end

    N = length(theta)

    l_theta_post = l_theta + N
    mu_theta_post = (sum(theta) + l_theta * m_theta) / l_theta_post
    mu_theta_new  = rand(Normal(mu_theta_post, sqrt(sigma2_theta / l_theta_post)))

    mu_theta = mu_theta_new

    a_theta_post = a_theta + (N + 1) / 2
    b_theta_post = b_theta + 0.5 * (sum((theta .- mu_theta).^2) + l_theta * (mu_theta - m_theta)^2)

    sigma2_theta_new = sample_from_inverse_gamma(a_theta_post, b_theta_post, seed)
    return mu_theta_new, sigma2_theta_new
end

# training loop

function train_fm_gibbs(
    model_params::Dict{Any, Any},
    latent_params::Dict{Any, Any},
    fixed_params::Dict{Any, Any},
    x_data::Array{Float64,2},
    y_data::Array{Float64,1},
    f_init::Array{Float64,1},
    num_iter::Int,
    seed::Union{Int, Nothing}=nothing
)
    loss_hist = []

    N = size(x_data, 2)
    K = size(model_params["v"], 2)
    D = size(x_data, 1)

    f = f_init
    q = calc_q_init(x_data, model_params["v"])

    for iter in 1:num_iter
        # noise parameter
        x_b, y_b = calc_xy_b(f, model_params["b"], x_data, y_data)
        latent_params["sigma2_noise"] = sample_posterior_sigma2_noise(x_b, y_b, model_params["b"], fixed_params["a_noise"], fixed_params["b_noise"])

        # hyperparameters
        latent_params["mu_w"], latent_params["sigma2_w"] = sample_posterior_mu_theta_and_sigma2_theta(model_params["w"], latent_params["sigma2_w"], fixed_params["m_w"], fixed_params["l_w"], fixed_params["a_w"], fixed_params["b_w"])
        for k in 1:K
            latent_params["mu_v"][k], latent_params["sigma2_v"][k] = sample_posterior_mu_theta_and_sigma2_theta(model_params["v"][:, k], latent_params["sigma2_v"][k], fixed_params["m_v"], fixed_params["l_v"], fixed_params["a_v"], fixed_params["b_v"])
        end

        # b
        x_b, y_b = calc_xy_b(f, model_params["b"], x_data, y_data)
        b_new    = sample_posterior_theta(x_b, y_b, latent_params["sigma2_noise"], fixed_params["mu_b"], fixed_params["sigma2_b"])
        f        = f + calc_df(x_b, b_new, model_params["b"])
        model_params["b"] = b_new

        # w
        for i in 1:N
            x_w, y_w = calc_xy_w(f, model_params["w"], x_data, y_data, i)
            w_i_new  = sample_posterior_theta(x_w, y_w, latent_params["sigma2_noise"], latent_params["mu_w"], latent_params["sigma2_w"])
            f        = f + calc_df(x_w, w_i_new, model_params["w"][i])
            model_params["w"][i] = w_i_new
        end

        # v
        for i in 1:N
            for k in 1:K
                x_v, y_v = calc_xy_v(f, q, model_params["v"], x_data, y_data, i, k)
                v_ik_new = sample_posterior_theta(x_v, y_v, latent_params["sigma2_noise"], latent_params["mu_v"][k], latent_params["sigma2_v"][k])
                f      = f      + calc_df(x_v, v_ik_new, model_params["v"][i, k])
                q[:,k] = q[:,k] + calc_dq(i, x_data, v_ik_new, model_params["v"][i, k])
                model_params["v"][i, k] = v_ik_new
            end
        end

        if (iter+1) % 100 == 0
            println("iter: $(iter+1), loss: $(sum((y_data - f) .^ 2) / D)")
        end
        push!(loss_hist, sum((y_data - f) .^ 2) / D)
    end

    return model_params, loss_hist
end

File: 202403/tests/FM_03_gibbs_julia.py
import numpy as np
from pathlib import Path
from typing import Optional
from julia import Main

current_dir = Path(__file__).resolve().parent
julia_file = current_dir / "FM_03_gibbs_julia.jl"
Main.include(str(julia_file))

# Factorization Machine

class FactorizationMachines:
    def __init__(self,
        num_features: int,
        num_factors:  int,
        sigma_b_init: float=0.,
        sigma_w_init: float=1.,
        sigma_v_init: float=1.,
        seed: Optional[int]=None
    ) -> None:
        self.rng = np.random.default_rng(seed)
        b = self.rng.normal(0, sigma_b_init)
        w = self.rng.normal(0, sigma_w_init, num_features)
        v = self.rng.normal(0, sigma_v_init, (num_features, num_factors))
        self.params = {'b': b, 'w': w, 'v': v}

    def predict(self, x: np.ndarray) -> float:
        if x.ndim == 1:
            x = x.reshape(1, -1) # x: (d, n)
        b = self.params['b']     # b: (1)
        w = self.params['w']     # w: (d)
        v = self.params['v']     # v: (d, k)

        bias   = b
            # (1)
        linear = x[:, :] @ w[:]
            # (D, N) @ (N) = (D)
        inter  = 0.5 * np.sum((x[:, :] @ v[:, :]) ** 2 - (x[:, :] ** 2) @ (v[:, :] ** 2), axis=1)
            # (D, K) -> (D)

        result = bias + linear + inter
            # (D)

        if result.shape[0] == 1:
            return float(result[0])
        return result

# test
N = 16
K = 8
D = 128

seed = 0
rng  = np.random.default_rng(seed)
fm   = FactorizationMachines(N, K, seed=seed)

Q      = rng.uniform(-1., 1., (N, N))
x_data = rng.choice((0, 1), size=(D, N))
y_data = np.einsum('dn,nm,dm->d', x_data, Q, x_data)

model_params = fm.params

latent_params = {
    'mu_w': 0.,
    'mu_v': np.zeros(K, dtype=float),
    'sigma2_w': 1.,
    'sigma2_v': np.ones(K, dtype=float)
}

fixed_params = {
    'mu_b': 0.,
    'sigma2_b': 1.,
    'm_w': 0.,
    'm_v': 0.,
    'l_w': 1.,
    'l_v': 1.,
    'a_w': 1.,
    'a_v': 1.,
    'b_w': 1.,
    'b_v': 1.,
    'a_noise': 1.,
    'b_noise': 1.,
}

_, loss_hist = Main.train_fm_gibbs(
    model_params,
    latent_params,
    fixed_params,
    x_data.astype(float),
    y_data,
    fm.predict(x_data),
    10000,
)
